### PR TITLE
Flintstone version 2.0

### PR DESCRIPTION
**Pull request by fire015**
* Major refactor, class names have changed and the whole codebase is much more extensible
* Removed the static `load` and `unload` methods
* The `replace` method is no longer public
* The `getFile` method has been removed
* Default swap memory limit has been increased to 2MB
* Ability to pass any instance for cache that implements `Doctrine\Common\Cache\Cache` and `Doctrine\Common\Cache\ClearableCache`

### CODE REVIEW

**Comment by fire015** [comment]
Hey @nyamsprod take a look at this and let me know thoughts.

**Comment by fire015** [comment]
CC @nikush

**Comment by nikush** [comment]
LGTM

**Comment by nyamsprod** [comment]
I have yet look into the codebase but why still supporting EOL PHP versions. Why not bump the min. require version to 5.5 ? In composer and elsewhere ?

**Comment by fire015** [comment]
I'd like to keep it backwards compatible for now, I don't see any fundamental issues with the code that would stop it working in older versions.

File: src/Config.php [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        'swap_memory_limit' => 2097152,    // 2MB
+    );
+
+    /**
+     * Constructor.
+     *
+     * @param array $config
+     */
+    public function __construct(array $config = array())
+    {
```
Comment (nyamsprod) [inline src/Config.php]: You should normalize the default `$config` array with `array_replace` or `array_merge` which would remove any need for the `if` this would include adding a `$defaultConfig` protected property IMHO

**Comment by nyamsprod** [comment]
If I've understand it correctly
- the `Database` object depends on the `Config` object
- the `Flintstone object depends on the `Database` and the `Config` object 

Why not simplify things by making the `Flinstone` depending only on the Database object ?

**Comment by nyamsprod** [comment]
IMHO the `Database` object exposes the wrong methods. Right now, you are only exposing  implementation details.

For example:
```php
        // Fetch the key from database
        $filePointer = $this->getDatabase()->openFile(Database::FILE_READ);
        $data = false;
        foreach ($filePointer as $line) {
            $data = $this->getDataFromLine($line, $key);
            if ($data !== false) {
                $data = $this->decodeData($data);
                break;
            }
        }
        $this->getDatabase()->closeFile($filePointer);
```
Should be written like this
```php
  $data = $this->getDatabase()->fetch($key);
```


TL;DR: You should have a `DatabaseInterface` and at least on concrete implementation with `SplFileObject`. This  interface should be fairly simple with the following methods

- has/contains($key)
- fetch($key);
- set($key);
- delete/remove($key);
- fetchAll();
- clear();

This way:

- I'll be able to switch Database backend and use for instance SQLite if I want to.
- Your coupling with `Doctrine` also will be more simple. 

Looking at your present code I fail to see the advantage of adding the `Database` object If my `Config` object already uses Doctrine ?

Hope my remarks make sense :)

File: src/Config.php (hunk 1 above)
Comment (fire015) [inline src/Config.php]: The reason it's done this way is so that each config that is provided run's through it's own setter method which does the validation. The protected $config contains the default value's if nothing is passed through the constructor.

Comment (nyamsprod) [inline src/Config.php]: I Understand that but normalizing the $config array will ensure:
- that all keys are present everytime
- that you can safely remove the ifs 
I did not say that you should remove the call the the set* methods :+1:

Comment (nyamsprod) [inline src/Config.php]: something like this is way more simpler IMHO
```php

public function __construct(array $config = array())
{
    $config = $this->normalizeConfig($config);
    $this->setDir($config['dir']);
    $this->setExt($config['ext']);
    $this->setGzip($config['gzip']);
    $this->setCache($config['cache']);
    $this->setFormatter($config['formatter']);
    $this->setSwapMemoryLimit($config['swap_memory_limit']);
}

protected normalizeConfig(array $config)
{
    $defaultConfig = array(
        'dir' => '',
        'ext' => '.dat',
        'gzip' => false,
        'cache' => true,
        'formatter' => null,
        'swap_memory_limit' => 2097152,    // 2MB
    );

    return array_replace($defautConfig, $config);
}
```

Comment (fire015) [inline src/Config.php]: Ok not a bad idea will update.

**Comment by fire015** [comment]
I have some concerns about this. The whole idea of Flintstone is that it's good at storing key/value data in a single file with locking, caching and gzip compression. If I start interfacing the Database class so that it can be swapped out, it just turns into a generic key/value data store class - which there are lots of packages for already. If you want to swap it for SQLite then Flintstone is not the right tool in the first place.

**Comment by nyamsprod** [comment]
Even if you don't want to create a interface or enable swapping of database... you are still coupling to much.  Does your `Flinstone` object really needs to know about how `Database` interact with your cache file ? Locking, Caching and Gzip are property of the `Database` not `Flinstone`. The `Flinstone` object simply does not care about it. It only cares about retrieving and adding data to the cache. That what `Database` should offer, and that what the API I suggest does. How `Database` act is solely dictated by the `Config` class. Which boilsdown to:

```php
$config = new Config($configArray);
$database = new Database($config);
$flinstone = new Flinstone($database);
```

my 0.02 cent.

PS: In this regards adding the Doctrine cache in the view seems odd as its add another caching layer to an already present cache.

**Comment by fire015** [comment]
That is true, I will update further based on those suggestions. Perhaps adding the static load method back in makes sense so that the Flintstone constructor only accepts a Database object but if you don't want the overhead of creating 3 objects you can call the load method? This would also keep it backwards compatible.

```php
public function __construct(Database $database) {
    $this->setDatabase($database);
}

public static function load($database, $config) {
    return new static(new Database($database), new Config($config));
}
```

**Comment by fire015** [comment]
Hey @nyamsprod please see latest changes. I have decided to drop doctrine as we only need a simple caching interface and I think it will confuse people if they see doctrine as a package.

**Comment by nyamsprod** [comment]
@fire015 Hi ..I've looked into your code and I have even submitted a PR with some improvement I still have some issue with Flintstone using a two layer cache.
- If Flintstone is too slow, then I won't use it and I will use another cache solution
- If Flintstone is not slow, then this extra cache layer is not needed

Another problem that I have is that I fail to see any distinction between the Database class and the Flintstone class ? Since the latter seems to be a proxy of the former ?

I've done some coding I'll submit to you if you are interested... probably tomorrow using a PR and tell me what you think about it. I've tried to resolve my two problems that I have with the current code.

**Comment by fire015** [comment]
Hi @nyamsprod I am totally confused now, you have an open PR with a refactor and then another refactor in your branch. I feel like the work I did already keeps it backwards compatible and simple which is why I am going ahead with it.

It is obvious there are many design patterns and different ways of implementing things as you have demonstrated, but as I say for backwards compatibility and simplicity I am going to stick to what I originally wrote. Thank you :)

**Comment by nyamsprod** [comment]
@fire015 the current PR is to improve the Config class in your v2. So you can accept or reject it if you feel it departs too much from your intent.

The other work is on **my fork** and has yet to be submitted. So again you are free to disregard it too. So I don't see the confusion. That being said accepted or not my current issue still stands:

- If Flintstone is too slow, then another cache solution will be used which render adding any additional cache layer useless. In the current code when using the additional cache your main cache mechanism (the file lines) is barely used and work as a backup solution at best.

- I fail to see any distinction between the Database class and the Flintstone class ? Since the latter is simply a proxy of the former ? If there are no distinction between both objects then one must simply be removed.

**Comment by fire015** [comment]
* Flintstone has always had caching in it, reading from files can be slow as they get bigger in size
* I have committed further to put the code back to how it was - Database just contains the utility methods on the database you wish to use
